internal/cloud/aws/metadata.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+package aws
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/feature/ec2/imds"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/edgelesssys/constellation/v2/internal/cloud/metadata"
+	"github.com/edgelesssys/constellation/v2/internal/role"
+)
+
+const (
+	tagName = "Name"
+	tagRole = "constellation-role"
+	tagUID  = "constellation-uid"
+)
+
+type ec2API interface {
+	DescribeInstances(context.Context, *ec2.DescribeInstancesInput, ...func(*ec2.Options)) (*ec2.DescribeInstancesOutput, error)
+}
+
+type imdsAPI interface {
+	GetInstanceIdentityDocument(context.Context, *imds.GetInstanceIdentityDocumentInput, ...func(*imds.Options)) (*imds.GetInstanceIdentityDocumentOutput, error)
+	GetMetadata(context.Context, *imds.GetMetadataInput, ...func(*imds.Options)) (*imds.GetMetadataOutput, error)
+}
+
+// Metadata implements core.ProviderMetadata interface for AWS.
+type Metadata struct {
+	ec2  ec2API
+	imds imdsAPI
+}
+
+func New(cfg aws.Config) *Metadata {
```

Comment:
All top-level, exported names should have doc comments, as should non-trivial unexported type or function declarations. [:link: Go Code Review Comments](https://github.com/golang/go/wiki/CodeReviewComments#doc-comments)

Reply:
If possible, don't expose structs from aws sdk outside of this package so other packages don't need to depend on int.

Reply:
Exactly :)